pkg/applicationserver/device_fetcher.go

```diff
@@ -174,15 +175,39 @@ func (f *circuitBreakerEndDeviceFetcher) circuitOpen() error {
 	return errCircuitBreakerOpen.New()
 }
 
-func (f *circuitBreakerEndDeviceFetcher) observeError(err error) {
+func (f *circuitBreakerEndDeviceFetcher) observeError(ctx context.Context, err error) {
+	infrastructureError := errors.IsCanceled(err) ||
+		errors.IsDeadlineExceeded(err) ||
+		errors.IsAborted(err) ||
+		errors.IsInternal(err) ||
+		errors.IsUnavailable(err)
+	logger := log.FromContext(ctx).WithField("circuit", "end_device_fetcher")
 	f.mu.Lock()
 	defer f.mu.Unlock()
-	if err != nil {
-		f.lastFailedAttempt = time.Now()
-		f.failures++
-	} else {
-		f.lastFailedAttempt = time.Time{}
-		f.failures = 0
+	switch {
+	case infrastructureError:
```

Comment:
What's the benefit of the var?

Reply:
The error code check is done outside the mutex, to minimize retention. I can't tell if this is premature or not.